### ISSUE TITLE
When pinning column, the grid menu shifts to the right of frozen column. The Grid Menu should stay at its original position.

### ISSUE THREAD

**Issue by DikshantBhatia** [ghiscoding/Angular-Slickgrid#400]
### Your Environment

| Software                | Version(s) |
| -------------------- | -------- |
| Angular                  |  7.2.0      | 
| Angular-Slickgrid   | 2.17.4    | 
| TypeScript              | 3.2.2       | 

### Steps to Reproduce
I create a custom column menu to pin columns to the left.  On pinning of column, the grid menu shifts from its position to the right of Frozen Column

### Expected Behavior
The Grid column should stay at its original position and should not be affected by frozen column or there should be a gridmenu option through we can fix its position.

### Current Behavior
The grid menu changes its position.

### Possible Solution

There is a code in slickgrid.menu.js which checks if there is a frozen column and based on that it decides the position of the the grid menu. We can have some custom property 'isPositionFixed' in gridMenu and I can set it to true. So, in slickgrid.menu.js, we only change the position of gridMenu if this property 'isPositionFixed' is false.

**Comment by ghiscoding** [comment]
That seems like a duplicate of #117 and as far as I know, the grid menu always stay in it's position in the [Example 20 - Frozen Grid](https://ghiscoding.github.io/Angular-Slickgrid/#/frozen)

Your issue doesn't have much to work with, you didn't provide any code neither print screen nor animated gif of your problem. I don't really understand your problem since again the example works just fine. I'm not good at guessing people's mind, so you're gonna have to provide much more than that if you want help.

![ClwY2u3KLT](https://user-images.githubusercontent.com/643976/74255395-ce30f180-4cbf-11ea-9e12-773906536815.gif)

**Comment by DikshantBhatia** [comment]
Okay ! I can give more information. Below is the code:

`app.component.ts
import {Component, OnInit} from '@angular/core';
import {AngularGridInstance, Column, Formatter, GridOdataService, GridOption, HeaderButtonItem, MenuCommandItem} from 'angular-slickgrid';

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.css']
})
export class AppComponent implements OnInit {

  title = 'AngularGridTest';
  angularGrid: AngularGridInstance;
  gridObj: any;
  columnDefinitions: Column[] = [];
  gridOptions: GridOption = {};
  menuCommandItems: MenuCommandItem[] = [];
  data: any[] = [];

  ngOnInit(): void {
    this.prepareGrid();
  }
  onGridReady(angularGrid: AngularGridInstance) {
    this.angularGrid = angularGrid;
    this.gridObj = angularGrid && angularGrid.slickGrid || {};
    console.log(this.gridObj);
  }
  prepareGrid() {
    this.gridOptions = {
        enableSorting :  true,
        enableColumnReorder: true,
        enableAutoSizeColumns : true,
        enableFiltering : true,
        enableExport: true,
        enableRowSelection : true,
        enableCheckboxSelector: true,
        checkboxSelector : {
          hideInColumnTitleRow : true,
          hideInFilterHeaderRow : false
        },
        rowSelectionOptions : {
          selectActiveRow : false
        },
        gridMenu : {
          hideForceFitButton : true,
        },
        headerMenu : {
          onCommand : (e, args) => {
              if (args.command === 'pinColumn') {
                console.log('current column: ' + args.column);

                const currentFrozenColumn =  this.gridObj.getOptions().frozenColumn;
                const currentColIndex =   this.gridObj.getColumnIndex(args.column.id);
                const cols: Column[]  = this.gridObj.getColumns();

                const currentColumn: Column = cols[currentColIndex];

                cols.splice(currentColIndex, 1);

                if (currentFrozenColumn === -1) {
                  cols.splice(1, 0 , currentColumn);
                } else {
                  cols.splice(currentFrozenColumn + 1, 0 , currentColumn);
                }

                /*this.angularGrid.gridStateService.resetColumns(cols);*/

                this.gridObj.setOptions({
                    frozenColumn : currentFrozenColumn === -1 ? 1 : currentFrozenColumn + 1
                  });

              }

          }
        },
    };

    // create a custom Formatter with the Formatter type
    // tslint:disable-next-line:max-line-length
    const myCustomHyperlinkFormatter: Formatter = (row: number, cell: number, value: any, 
 columnDef: Column, dataContext: any, grid?: any) =>
      `<button class="btn btn-link"  aria-hidden="true">${value}</button>`;

    this.menuCommandItems = [
      {command : 'pinColumn', title : 'Pin to Left'}
    ];

    this.columnDefinitions = [
      // tslint:disable-next-line:max-line-length
      {id: 'source', field: 'source', name: 'Source', sortable : true, filterable: true, formatter: myCustomHyperlinkFormatter, cssClass : 'sourceColumnCell', onCellClick : function(e , args) {
          console.log(args);
        } },
      {id: 'hostname', field: 'hostname', name: 'Host Name', sortable :  true, filterable: true},
      {id: 'port', field: 'port', name: 'Port Number', sortable :  true, filterable: true, header : {menu : {items : this.menuCommandItems}}},
      // tslint:disable-next-line:max-line-length
      {id: 'remoteFolder', field: 'remoteFolder', name: 'Remote Folder', sortable : true, filterable: true, header : {menu : {items : this.menuCommandItems}}}
    ];

    this.prepareData();
  }

  prepareData() {
    this.data = [{ id: '1', source: 'dccm', hostname : 'localhost', port : '8080',  remoteFolder : 'tmp'},
      { id: '2' , source: 'encts', hostname : 'localhost', port : '8000',  remoteFolder : 'dir'},
      { id: '3' , source: 'pcm', hostname : 'localhost', port : '4000',  remoteFolder : 'remote/tmp'},
      { id: '4' , source: 'rid', hostname : 'localhost', port : '1000',  remoteFolder : 's:/d'},
      { id: '5' , source: 'aim', hostname : 'localhost', port : '2000',  remoteFolder : 'users/dbhatia'},
      { id: '6' , source: 'pam', hostname : 'localhost', port : '4000',  remoteFolder : 'tmp'},
      { id: '7' , source: 'dab', hostname : 'localhost', port : '9000',  remoteFolder : 'tmp'}
    ];
  }
  onSelectedRowsChanged(eventData, args) {
        args.rows.map(row => {
        const item = this.gridObj.getDataItem(row);
        console.log(item.source);
      });
  }
}
`
app.component.html
`<div class="container">
  <angular-slickgrid gridId="grid1"
                     [columnDefinitions]="columnDefinitions"
                     [gridOptions]="gridOptions"
                     [dataset]="data"
                     (onAngularGridCreated)="onGridReady($event)"
                      (sgOnSelectedRowsChanged)= "onSelectedRowsChanged($event.detail.data, $event.detail.args)">
  </angular-slickgrid>
</div>
`



![GridMenuMovingOnPinning](https://user-images.githubusercontent.com/9398275/74335008-7798e580-4d9b-11ea-93dc-5b774961d20a.gif)

**Comment by ghiscoding** [comment]
You're bypassing the lib internal code and that is most probably why it doesn't work. Internally, the lib keeps a reference to all the columns (and all visible columns), so that if you go and hide a column, the lib knows which one you've hide and it could be reset on demand... but in your case, you're most probably bypassing that completely and then the lib doesn't know that you moved a column. You're not supposed to play directly with the column definitions.

You can maybe do these 2 things

1. Try moving it manually by doing a drag & drop of the column to first position, then only after do the pinning. 
2. The other thing you could maybe try is to at least advise the lib that you've changed the column definitions, you can update the column definitions object so that the lib gets informed
```ts
this.columnDefinitions = cols.slice();
```
A good example of this is when user dynamically add a new column to the definitions, it has to be done like [this](https://github.com/ghiscoding/Angular-Slickgrid/blob/master/src/app/examples/grid-editor.component.ts#L591) for it to work. Which is basically to push to the column definitions array and then do a shallow copy of the array so that the lib knows you've made some changes and so it re-render and refreshes its internal arrays.

Basically if you're bypassing the lib, it's at your own risk and I won't help you more than that and actually this kind of question is better to ask on Stack Overflow since this is not a problem with the lib  itself.

**Comment by DikshantBhatia** [comment]
Thanks for the reply.
After updating column definition with the cols, the header menu stays at its position. But, ideally the cols are not column definition as they also have the checkbox column inside it. But, i can write the logic to deal this. 
I wonder why "this.angularGrid.gridStateService.resetColumns(cols)" did not do the trick. This means we have to update columnDefinitions everytime explicity.
Anyways, thanks a lot or the help .

**Comment by ghiscoding** [comment]
> I wonder why "this.angularGrid.gridStateService.resetColumns(cols)" did not do the trick. This means we have to update columnDefinitions everytime explicity.

It's not meant to be used that way, this is a reset of the columns to its original declaration and is used in this [Example 16 - Grid State & Presets](https://ghiscoding.github.io/Angular-Slickgrid/#/gridstate) to Reset the grid to its original column definitions BUT what you are looking for is a Set. So the best way to do a **Set** is like I told you earlier. 

> But, ideally the `cols` are not column definition as they also have the checkbox column inside it.

In your case they are the column definitions, because you called `const cols: Column[]  = this.gridObj.getColumns();` and when you execute `getColumns()` it is the method that SlickGrid uses to get the column definitions array.

> This means we have to update columnDefinitions everytime explicity.

Yes that is the only way to make sure the lib is in sync with the reality. You can use all of the SlickGrid (core) methods but when you do, you have to make sure that Angular-Slickgrid knows of the changes to you applied directly to the core, the best way that I've found was to expose that `columnDefinitions` Setter so that the user (you) can override the column definitions. What could possibly be done is to add another method in maybe the `GridService` to do this set columns, the code would basically do the same thing as the column definitions Setter, so I'm not sure what benefit that could bring.

BTW to include some code in markdown on GitHub you can use the triple backtick then the language to use and when you're done close the 3 backticks. ` ```ts ` or ` ```html ` and that makes it much easier to read 😃 

I'll close the issue since I believe I answered all your questions and you now know what to do.

Cheers ⭐️

**Comment by ghiscoding** [comment]
@DikshantBhatia 
I'm looking at adding this as a built-in feature to pin column to the left of current column. What code did you end up using for this to work? Isn't just to find the column index in the column definition array and then change the `frozenColumn` with that column index position?

**Comment by DikshantBhatia** [comment]
@ghiscoding 
At that time, the thing which i was trying to achieve was to have the pinned columns in a container on left hand side. So, when you pin the column to the left, it goes to the left container(right of the last column in the left container). Below code was used to achieve that:

```
if (args.command === 'pinColumn') {
                console.log('current column: ' + args.column);

                const currentFrozenColumn =  this.gridObj.getOptions().frozenColumn;
                const currentColIndex =   this.gridObj.getColumnIndex(args.column.id);
                const cols: Column[]  = this.gridObj.getColumns();

                const currentColumn: Column = cols[currentColIndex];

                cols.splice(currentColIndex, 1);

                if (currentFrozenColumn === -1) {
                  cols.splice(1, 0 , currentColumn);
                } else {
                  cols.splice(currentFrozenColumn + 1, 0 , currentColumn);
                }

                /*this.angularGrid.gridStateService.resetColumns(cols);*/
                this.columnDefinitions = cols;

                this.gridObj.setOptions({
                    frozenColumn : currentFrozenColumn === -1 ? 1 : currentFrozenColumn + 1
                  });

              }
```

**Comment by ghiscoding** [comment]
Just wanted to mention that this feature (Freeze Columns) I talked earlier is now available and released, more info [releases](https://github.com/ghiscoding/Angular-Slickgrid/releases). However note it's an opt-in feature

**Comment by DikshantBhatia** [comment]
@ghiscoding  It looks good. Thank you ! I will have a look at it :)

**Comment by Vinoth-deva** [comment]
> Just wanted to mention that this feature (Freeze Columns) I talked earlier is now available and released, more info [releases](https://github.com/ghiscoding/Angular-Slickgrid/releases). However note it's an opt-in feature. Thanks for the idea.

is it possible to pin the right side column???

**Comment by ghiscoding** [comment]
No... only the left

**Comment by ranjithitwizard** [comment]
@ghiscoding 

Hi, I found an issue, when you hide/show the pinned columns(left side columns) from the grid, the unpinned columns(right side columns) moved to pinned columns rather than stick to the unpinned one.. any possibility to prevent this action ?

Lets say;

  Pinned columns    | Unpinned columns
**Col1 | Col2 | Col3** | Col4  | Col5 | Col6 

If **Col2** is removed the from the grid, then **Col4** is moving towards to pinned columns section

Pinned columns   | Unpinned columns
**Col1 | Col3  | Col4**  | Col5 | Col6 

I want to stick with only **Col1 & Col3** to be pinned not the unpinned columns.. any workaround for this issue ?

**Comment by ghiscoding** [comment]
That's because SlickGrid pinning works by index `frozenColumn: number` and it has no knowledge whatsoever of the columns themselves. By taking a quick look at the code, I see that it will require multiple changes for this to work
1. when **hiding** a column, we need to see if we are using `frozenColumn` and if we are then we also need to detect if the new hidden column is on the left container, if so then re-adjust `frozenColumn` (doing a -1 on the index should suffice)
2. step 1 covers the hiding but we also need to cover the **showing** back of a column from any of the pickers (columnPicker/gridMenu), I'm assuming that can be done by listening on `onColumnsChanged` event on both and doing a reverses analysis of step 1 

The only workaround is for the user to re-adjust the pinning manually via the header menu... unless you want to implement those steps in your own grid

You should open a separate issue since this is **totally unrelated** to this issue, so please don't pollute issues that have already been fixed.

**Comment by ghiscoding** [comment]
@ranjithitwizard 
The problem you mentioned is now fixed via PR #642 under version [2.23.3](https://github.com/ghiscoding/Angular-Slickgrid/releases/tag/v2.23.3). 

The issue fix is covering the following 3 scenarios (which internally will recalculate the `frozenColumn` index position every time the following action is executed):
1. hide a column shouldn't affect pinning on left container
2. hiding/showing a column via Column Picker shouldn't change pinning on left container
3. hiding/showing a column via Grid Menu shouldn't change pinning on left container

I'll say it again though, please create a **new issue** instead of commenting on old issue
Cheers ⭐
